File: app/src/main/java/com/example/insta/AddImage.java
package com.example.insta;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.Timestamp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.Blob;
import com.google.firebase.firestore.FirebaseFirestore;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Date;
import java.util.HashMap;
import java.util.Objects;

import Controller.Image;
import Model.Post;

public class AddImage extends Fragment {

    public static final int ADD_IMAGE_REQUEST = 1025;

    ImageView image;
    EditText caption;
    Button add;

    FirebaseAuth auth;
    Uri uri;

    private ProgressBar progressBar;
    private View view;
    private Bitmap reducedBitmap;


    public AddImage() {
        auth = FirebaseAuth.getInstance();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.activity_add_image, container, false);

        image = view.findViewById(R.id.image_add);
        caption = view.findViewById(R.id.caption);
        add = view.findViewById(R.id.add_add_image);
        progressBar = view.findViewById(R.id.progress_bar);


        image.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                addImageClicked(v);
            }
        });

        add.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                addClicked(v);
            }
        });

        return view;
    }

    public void addImageClicked(View view) {
        image.setEnabled(false);
        Intent intent = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        startActivityForResult(intent, ADD_IMAGE_REQUEST);
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        image.setEnabled(true);
        super.onActivityResult(requestCode, resultCode, data);

        if (data != null && requestCode == ADD_IMAGE_REQUEST && resultCode == -1) {
            uri = data.getData();
            if (uri != null) {
                try {
                    Bitmap bitmap = MediaStore.Images.Media.getBitmap(getActivity().getContentResolver(), uri);
                    reducedBitmap = Image.getReducedBitmap(bitmap, 512);
                    Log.d("Image Uploaded", String.valueOf(reducedBitmap.getByteCount()));
                    image.setImageBitmap(bitmap);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public void addClicked(View view) {
        if (reducedBitmap != null) {
            image.setEnabled(false);
            add.setEnabled(false);
            caption.setEnabled(false);
        } else {
            Toast.makeText(view.getContext(), "Add an image firstly", Toast.LENGTH_LONG).show();
            return;
        }

        progressBar.setVisibility(View.VISIBLE);

        Post post = new Post();
        post.setCaption(caption.getText().toString());
        post.setTime(Timestamp.now());
        post.setLikes(new HashMap());
        post.setComments(new HashMap());
        post.setShares(new HashMap());
        post.setLiked(false);
        post.setUID(auth.getUid());
        post.setID(String.valueOf(Timestamp.now().getSeconds()));

        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        reducedBitmap.compress(Bitmap.CompressFormat.PNG, 0, stream);
        byte[] byteArray = stream.toByteArray();
        post.setPostImage(Blob.fromBytes(byteArray));

        Log.d("byteArray" , byteArray.length+"");

        FirebaseFirestore.getInstance()
                .collection("Users")
                .document(Objects.requireNonNull(auth.getUid()))
                .collection("Posts")
                .document(post.getID()).set(post).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    Toast.makeText(getContext(), "The image added Successfully", Toast.LENGTH_LONG).show();
                } else {
                    Toast.makeText(getContext(), "Failed to upload the image", Toast.LENGTH_LONG).show();
                }

                caption.setText("");
                image.setImageResource(R.drawable.add_image);
                image.setEnabled(true);
                add.setEnabled(true);
                reducedBitmap = null;
                caption.setEnabled(true);
                progressBar.setVisibility(View.GONE);
            }
        });
    }
}

File: app/src/main/java/com/whiteside/insta/AddImage.kt
package com.whiteside.insta

import Controller.Image
import Model.Post
import android.Manifest
import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.google.android.gms.tasks.Task
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.Blob
import com.google.firebase.firestore.FirebaseFirestore
import com.whiteside.insta.databinding.ActivityAddImageBinding
import java.io.ByteArrayOutputStream
import java.io.IOException

class AddImage : Fragment() {
    private val auth: FirebaseAuth = FirebaseAuth.getInstance()
    private var bind: ActivityAddImageBinding? = null
    private var reducedBitmap: Bitmap? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {

        bind = ActivityAddImageBinding.inflate(layoutInflater)
        bind!!.addImage.setOnClickListener { addImageClicked() }
        bind!!.add.setOnClickListener { addClicked() }

        return bind!!.root
    }

    private fun verifyPermission() {
        if (requireActivity().checkSelfPermission(Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            startActivityForResult(intent, TAKE_PICTURE)
        } else {
            requireActivity().requestPermissions(arrayOf(Manifest.permission.CAMERA), PERMISSION_REQUEST_CODE)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == PERMISSION_REQUEST_CODE && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            verifyPermission()
        }
    }

    private fun addImageClicked() {
        bind!!.addImage.isEnabled = false
        AlertDialog.Builder(context)
                .setTitle("Select photo way")
                .setPositiveButton("Take a picture") { dialog: DialogInterface?, which: Int -> verifyPermission() }
                .setNegativeButton("Select a picture") { dialog: DialogInterface?, which: Int ->
                    val intent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
                    startActivityForResult(intent, ADD_IMAGE_REQUEST)
                }
                .setOnDismissListener { dialog: DialogInterface? -> bind!!.addImage.isEnabled = true }
                .show()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        bind!!.add.isEnabled = true
        bind!!.caption.isEnabled = true
        bind!!.addImage.isEnabled = true
        if (data != null && requestCode == ADD_IMAGE_REQUEST && resultCode == -1) {
            val uri = data.data
            if (uri != null) {
                try {
                    val bitmap = MediaStore.Images.Media.getBitmap(activity?.contentResolver, uri)
                    reducedBitmap = Image.getReducedBitmap(bitmap, 512)
                    Log.d("Image Uploaded", reducedBitmap!!.byteCount.toString())
                    bind!!.addImage.setImageBitmap(bitmap)
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        } else if (requestCode == TAKE_PICTURE && resultCode == -1) {
            val bitmap = data!!.extras!!["data"] as Bitmap?
            reducedBitmap = bitmap
            bind!!.addImage.setImageBitmap(bitmap)
        }
    }

    fun addClicked() {
        if (reducedBitmap != null) {
            bind!!.add.isEnabled = false
            bind!!.caption.isEnabled = false
            bind!!.addImage.isEnabled = false
        } else {
            Toast.makeText(requireContext(), "Add an image firstly", Toast.LENGTH_LONG).show()
            return
        }
        bind!!.progressBar.visibility = View.VISIBLE
        val post = Post(
                caption = bind!!.caption.text.toString(),
                time = Timestamp.now(),
                uId = auth.uid,
                Id = Timestamp.now().seconds.toString()
        )
        val stream = ByteArrayOutputStream()
        reducedBitmap!!.compress(Bitmap.CompressFormat.PNG, 100, stream)
        val byteArray = stream.toByteArray()
        post.postImage = Blob.fromBytes(byteArray)
        Log.d("byteArray", byteArray.size.toString() + "")
        FirebaseFirestore.getInstance()
                .collection("Users")
                .document(auth.currentUser!!.uid)
                .collection("Posts")
                .document(post.Id!!)
                .set(post)
                .addOnCompleteListener { task: Task<Void?> ->
                    if (task.isSuccessful) {
                        Toast.makeText(context, "The image added Successfully", Toast.LENGTH_LONG).show()
                    } else {
                        Toast.makeText(context, "Failed to upload the image", Toast.LENGTH_LONG).show()
                    }
                    bind!!.caption.setText("")
                    bind!!.addImage.setImageResource(R.drawable.add_image)
                    bind!!.addImage.isEnabled = true
                    bind!!.add.isEnabled = true
                    reducedBitmap = null
                    bind!!.caption.isEnabled = true
                    bind!!.progressBar.visibility = View.GONE
                }
    }

    companion object {
        const val ADD_IMAGE_REQUEST = 1025
        const val TAKE_PICTURE = 2015
        const val PERMISSION_REQUEST_CODE = 123
    }
}